### DISCUSSION TITLE
I'm facing this issue: could not determine kind of name for C.BPF_MAP_TYPE_BLOOM_FILTER

### DISCUSSION THREAD

**Discussion by pratikjagrut** [aquasecurity/libbpfgo#401] [Q&A] [answered]
Hi,

I'm diving into this library for the first time, and I'm pretty new when it comes to eBPF applications. While trying to build one of the apps I'm tinkering with using the `libbpfgo` library, I ran into this snag:

```
   ⨯ build failed after 2.17s error=failed to build for linux_amd64: exit status 1: # github.com/aquasecurity/libbpfgo
../../../../go/pkg/mod/github.com/aquasecurity/libbpfgo@v0.6.0-libbpf-1.3/map-common.go:51:39: could not determine kind of name for C.BPF_MAP_TYPE_BLOOM_FILTER
```

I've got this feeling it might be a versioning or compatibility thing on my end, but I'm a bit stumped. Could someone help me figuring this out?

```
bpftool --version

bpftool v7.3.0
using libbpf v1.3
features:

```

**Comment by geyslan** [comment]
Hey @pratikjagrut, you're see the version used by bpftool. Could you check the system installed one?

![image](https://github.com/aquasecurity/libbpfgo/assets/3372117/f189e70b-4fea-4473-aa21-0ef076e9ee46)

For more on how to use libbpfgo see https://github.com/aquasecurity/libbpfgo/issues/400#issuecomment-1909184964.

**Comment by pratikjagrut** [comment]
@geyslan I don't see libbpf_version.h file at that location.

**Reply by geyslan** [reply]
I mentioned the version file to pinpoint where you could check your system libbpf version. Anyway, if you're facing such error, it is certainly related to the lack of definition in your includes.

Please check your build process and the `-I` flag to find out where it points. If you're not using, it will try to include it from the system path. Mine has `BPF_MAP_TYPE_BLOOM_FILTER` defined.

```
❯ grep -ir 'BPF_MAP_TYPE_BLOOM_FILTER' /usr/include/linux 
/usr/include/linux/bpf.h:	BPF_MAP_TYPE_BLOOM_FILTER,
/usr/include/linux/bpf.h:		 * BPF_MAP_TYPE_BLOOM_FILTER - the lowest 4 bits indicate the
```

Anyhow, I would recommend you to follow the selftest building way (static compilation), since it uses the correct libbpf version downloaded with libbpfgo. It's kept locally, not installed to the system.

https://github.com/aquasecurity/libbpfgo/blob/90dbffffbdabd70bb8e834b8777bf12186e1ae8c/selftest/common/Makefile#L17-L18

https://github.com/aquasecurity/libbpfgo/blob/90dbffffbdabd70bb8e834b8777bf12186e1ae8c/selftest/common/Makefile#L57-L64

**Reply by pratikjagrut** [reply]
Our build target is configured as follows:

```makefile
BASEDIR = $(abspath ./)
OUTPUT = ./output
ARCH := $(shell uname -m | sed 's/x86_64/amd64/g; s/aarch64/arm64/g')

LIBBPF_SRC = $(abspath ./3rdparty/libbpf/src)
LIBBPF_OBJ = $(abspath $(OUTPUT)/libbpf.a)
LIBBPF_OBJDIR = $(abspath $(OUTPUT)/libbpf)
LIBBPF_DESTDIR = $(abspath $(OUTPUT))

CC = gcc
CLANG = clang
GO = go
CFLAGS = -g -O2 -Wall -fpie
LDFLAGS =

CGO_CFLAGS_STATIC = "-I$(abspath $(OUTPUT)) -Wno-unknown-attributes"
CGO_LDFLAGS_STATIC = "-lelf -lz $(LIBBPF_OBJ)"
CGO_EXTLDFLAGS_STATIC = '-w -extldflags "-static"'
CGO_CFGLAGS_DYN = "-I. -I/usr/include/"
CGO_LDFLAGS_DYN = "-lelf -lz -lbpf"
```
libbpf version is the latest one `commit 20c0a9e3d7e7d4aeb283eae982543c9cacc29477 (HEAD, tag: v1.3.0)`
When executed, it generates the `output/vmlinux.h` file containing the definition for `BPF_MAP_TYPE_BLOOM_FILTER`.

**Reply by geyslan** [reply]
The tracee building might help you identify what's wrong:

https://github.com/aquasecurity/tracee/blob/main/Makefile#L314-L371

**Reply by ceeeeb-hao** [reply]
Why is there no output when I run the command: "grep -ir 'BPF_MAP_TYPE_BLOOM_FILTER' /usr/include/linux"?

**Reply by rscampos** [reply]
tks for you msg @ceeeeb-hao 
one possibility is that your kernel don't have this const BPF_MAP_TYPE_BLOOM_FILTER defined. According with this [ref](https://docs.kernel.org/bpf/map_bloom_filter.html) it was introduced in kernel 5.16.

**Reply by ceeeeb-hao** [reply]
I also looked at this reference, but my kernel is 6.5

uname -srm
Linux 6.5.0-35-generic x86_64

**Reply by geyslan** [reply]
It should be available.

- https://elixir.bootlin.com/linux/v6.5/source/include/uapi/linux/bpf.h#L943
- https://github.com/libbpf/libbpf/blob/master/include/uapi/linux/bpf.h#L1009

But anyway, do not rely on the system libbpf, since it can mismatch the one used by libbpfgo. Try the approach of including the one available in the libbpf uapi (provided by libbpfgo), as @rscampos [suggested](https://github.com/aquasecurity/libbpfgo/discussions/401#discussioncomment-10084893).

Let us know. Tks for using libbpfgo.

**Comment by ceeeeb-hao** [comment]
I am also experiencing your problem, have you found a solution?
![报错](https://github.com/user-attachments/assets/cbc6ba02-5179-4388-94d9-b3be8f5d6393)

ubuntu22.04

**Reply by rscampos** [reply]
Some days ago we add this [commit](https://github.com/aquasecurity/libbpfgo/commit/c8f76d124d7dd4aff1fec4d5a2a738b811381527) in our main branch, pls take a look. We add a full path (`./libbpf/include/uapi`) to be passed down to go build via `CGO_CFLAGS`. This way, even if the kernel does not have the symbol, libbpf does - as we can see in the [libbpf repo](https://github.com/libbpf/libbpf/blob/686f600bca59e107af4040d0838ca2b02c14ff50/include/uapi/linux/bpf.h#L1009).

 I'd like to ask you to test it and ping me back if work or not. Tks!

**Reply by ceeeeb-hao** [reply]
I'm very sorry to say that, having seen the power of tracee, this is the first time I've used libbpfgo to write an eBPF program! May I ask if you mean this:
CC=clang CGO_CFLAGS="-I /home/chen/eBPF/libbpfgo/output/linux" CGO_LDFLAGS="/usr/lib/x86_64-linux-gnu/libbpf.a" go build -o test

Tks！

**Reply by geyslan** [reply, marked as answer]
Take for instance the flags from tracee building `make tracee`:

`GOOS=linux CC=clang GOARCH=amd64 CGO_CFLAGS="-I/home/gg/code/tracee/dist/libbpf -I/home/gg/code/tracee/3rdparty/libbpf/include/uapi" CGO_LDFLAGS="-L/home/gg/code/tracee/dist/libbpf/obj -lbpf" go build`

For a static building `make tracee STATIC=1`:

`GOOS=linux CC=clang GOARCH=amd64 CGO_CFLAGS="-I/home/gg/code/tracee/dist/libbpf -I/home/gg/code/tracee/3rdparty/libbpf/include/uapi" CGO_LDFLAGS="-L/home/gg/code/tracee/dist/libbpf/obj -lbpf -lelf -lzstd -lz" go build`

Notice that I'm not pointing to the system's libbpf. I suggest you do the same.

For more, check https://github.com/aquasecurity/tracee/blob/main/Makefile#L470-L489

**Reply by tsint** [reply]
> Some days ago we add this [commit](https://github.com/aquasecurity/libbpfgo/commit/c8f76d124d7dd4aff1fec4d5a2a738b811381527) in our main branch, pls take a look. We add a full path (`./libbpf/include/uapi`) to be passed down to go build via `CGO_CFLAGS`. This way, even if the kernel does not have the symbol, libbpf does - as we can see in the [libbpf repo](https://github.com/libbpf/libbpf/blob/686f600bca59e107af4040d0838ca2b02c14ff50/include/uapi/linux/bpf.h#L1009).
> 
> I'd like to ask you to test it and ping me back if work or not. Tks!

I regret to inform you that this commit is completely ineffective because the libbpf submodule is not included in the go module cache where libbpfgo resides, so the same error will still occur.
```
go build -ldflags="-extldflags=-s -w" -o tracer .
# github.com/aquasecurity/libbpfgo
/home/dc/go/pkg/mod/github.com/aquasecurity/libbpfgo@v0.7.0-libbpf-1.4.0.20240715144951-c8f76d124d7d/map-common.go:51:39: could not determine kind of name for C.BPF_MAP_TYPE_BLOOM_FILTER
make: *** [Makefile:27: build] Error 1
dc@mint213:~/go/pkg/mod/github.com/aquasecurity/libbpfgo@v0.7.0-libbpf-1.4.0.20240715144951-c8f76d124d7d$ ls
btf.go         buf-perf.go  builder  elf.go  go.sum   helpers_test.go  libbpfgo.c   libbpfgo.h  link.go         logger_cb.go       Makefile       map.go           map-low.go  module.go           prog-common.go  Readme.md   rwArray_test.go  tchook-common.go
buf-common.go  buf-ring.go  docs     go.mod  go.work  libbpf_cb.go     libbpfgo.go  LICENSE     link-reader.go  logger_cb_test.go  map-common.go  map-iterator.go  misc.go     module-iterator.go  prog.go         rwArray.go  selftest         tchook.go

```

**Reply by geyslan** [reply]
@tsint please take a look in https://github.com/aquasecurity/libbpfgo/discussions/401#discussioncomment-10089843 above.

**Reply by tsint** [reply]
> @tsint please take a look in [#401 (reply in thread)](https://github.com/aquasecurity/libbpfgo/discussions/401#discussioncomment-10089843) above.

Thank you, that can be compiled normally, but it's still inconvenient.

**Reply by geyslan** [reply]
As libbpfgo is a wrapper of libbpf it embeds a specific version (required to make features to work). What you see as inconvenient is probably a requirement of that libbpf version itself. So as new versions are brought into libbpfgo with new features for newest kernels, some dependencies raise as well. So it is just the way it is.

As mentioned above by @rscampos, a fix was done regarding the includes that weren't set properly before (libbpfgo selftests). All flags are now generated from the embedded version. That's the way to follow to have a sane building, otherwise different includes or flags will be used (possibly from different version).

If you think in a better way to improve how libbpfgo tackles the wrapping and building for the consumer, please bring us. It will be awesome. Perhaps a PR with better instructions? Let us know.
